[Numberinput] Validation may not work with Vue 3.4.28 or higher

### Overview of the problem

**Buefy-next** version: [0.2.0]
**Vuejs** version: [3.5.13]
**OS/Browser**: macOS / Vitest

### Description

The spec for `Numberinput` fails with the following error:
```
 FAIL  src/components/numberinput/Numberinput.spec.ts > BNumberinput > Rendered > is invalid when step / minStep decimals and value decimals lengths are different
AssertionError: expected true to deeply equal false

- Expected
+ Received

- false
+ true

 ❯ src/components/numberinput/Numberinput.spec.ts:163:67
    161|         it('is invalid when step / minStep decimals and value decimals…
    162|             await wrapper.setProps({ step: 1, modelValue: 1.15 })
    163|             expect(wrapper.find('input').element.checkValidity()).toEq…
       |                                                                   ^
    164|             await wrapper.setProps({ step: 1.15, modelValue: 1.154 })
    165|             expect(wrapper.find('input').element.checkValidity()).toEq…
```

I found this happens with Vue v3.4.28 or higher. And a **user can type an invalid value in the `Numberinput` field without any validation error**.

### Steps to reproduce

Run the unit tests:
```sh
npm -w @ntohq/buefy-next run unit
```

### Expected behavior

The test passes.

### Actual behavior

The test fails.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -6,6 +6,7 @@
 
 * [#274](https://github.com/ntohq/buefy-next/issues/274) `Table`, and `Tabs` caused memory leaks on Vue 3.5.
   `SlotComponent` which is internally used by `Table`, and `Tabs` ceased adding the update hook to the component specified to the `component` prop, and dropped the `event` prop.
+* [#429](https://github.com/ntohq/buefy-next/issues/429) `Numberinput` may have accepted an invalid number as valid with Vue 3.4.28 or higher.
 
 ## Buefy-next 0.2.0
 

--- a/packages/buefy-next/src/components/input/Input.spec.ts
+++ b/packages/buefy-next/src/components/input/Input.spec.ts
@@ -229,23 +229,39 @@ describe('BInput', () => {
             spyOnCheckHtml5Validity.mockReset()
         })
 
-        it('should validate value when updated by user interaction', async () => {
+        it('should validate value at input event', async () => {
             const wrapper = shallowMount(BInput, {
                 data: () => ({ isValid: false } as BInputData)
             })
 
-            // simulates "input" event
-            // eslint-disable-next-line @typescript-eslint/no-explicit-any
-            wrapper.vm.onInput({ target: { value: 'foo' } } as any)
-            await wrapper.vm.$nextTick()
+            const inputElement = wrapper.get('input')
+
+            await inputElement.element.focus()
+            expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(0)
+
+            await inputElement.trigger('input')
             expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(1)
 
-            // simulates "change" event
-            await wrapper.setProps({ lazy: true })
-            // eslint-disable-next-line @typescript-eslint/no-explicit-any
-            wrapper.vm.onChange({ target: { value: 'bar' } } as any)
-            await wrapper.vm.$nextTick()
-            expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(2)
+            await inputElement.trigger('change')
+            expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(1)
+        })
+
+        it('should validate value at change event if lazy', async () => {
+            const wrapper = shallowMount(BInput, {
+                props: { lazy: true },
+                data: () => ({ isValid: false } as BInputData)
+            })
+
+            const inputElement = wrapper.get('input')
+
+            await inputElement.element.focus()
+            expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(0)
+
+            await inputElement.trigger('input')
+            expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(0)
+
+            await inputElement.trigger('change')
+            expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(1)
         })
 
         it('should validate value when programmatically updated', async () => {
@@ -280,21 +296,33 @@ describe('BInput', () => {
                 spyOnCheckHtml5Validity.mockClear()
             })
 
-            it('should validate value once when updated by user interaction', async () => {
-                // simulates "input" event
-                // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                wrapper.vm.onInput({ target: { value: 'foo' } } as any)
-                await wrapper.vm.$nextTick()
+            it('should update and validate value at input event', async () => {
+                const inputElement = wrapper.get('input')
+
+                inputElement.element.value = 'foo'
+
+                await inputElement.trigger('input')
                 expect(root.vm.value).toBe('foo')
                 expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(1)
 
-                // simulates "change" event
+                await inputElement.trigger('change')
+                expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(1)
+            })
+
+            it('should update and validate value at change event if lazy', async () => {
                 await root.setData({ lazy: true })
-                // eslint-disable-next-line @typescript-eslint/no-explicit-any
-                wrapper.vm.onChange({ target: { value: 'bar' } } as any)
-                await wrapper.vm.$nextTick()
-                expect(root.vm.value).toBe('bar')
-                expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(2)
+                // input element must be queried after changing `lazy`
+                const inputElement = root.get('input')
+
+                inputElement.element.value = 'foo'
+
+                await inputElement.trigger('input')
+                expect(root.vm.value).toBe('')
+                expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(0)
+
+                await inputElement.trigger('change')
+                expect(root.vm.value).toBe('foo')
+                expect(spyOnCheckHtml5Validity).toHaveBeenCalledTimes(1)
             })
 
             it('should validate value once when programmatically updated', async () => {

--- a/packages/buefy-next/src/components/input/Input.vue
+++ b/packages/buefy-next/src/components/input/Input.vue
@@ -4,35 +4,68 @@
         :class="rootClasses"
         v-bind="rootAttrs"
     >
-        <input
-            v-if="type !== 'textarea'"
-            ref="input"
-            class="input"
-            :class="[inputClasses, customClass]"
-            :type="newType"
-            :autocomplete="newAutocomplete"
-            :maxlength="maxlength"
-            :value="computedValue"
-            v-bind="fallthroughAttrs"
-            @input="onInput"
-            @change="onChange"
-            @blur="onBlur"
-            @focus="onFocus"
-        >
+        <template v-if="lazy">
+            <input
+                v-if="type !== 'textarea'"
+                ref="input"
+                class="input"
+                :class="[inputClasses, customClass]"
+                :type="newType"
+                :autocomplete="newAutocomplete"
+                :maxlength="maxlength"
+                v-model.lazy="computedValue"
+                v-bind="fallthroughAttrs"
+                @input="onInput"
+                @change="onChange"
+                @blur="onBlur"
+                @focus="onFocus"
+            >
 
-        <textarea
-            v-else
-            ref="textarea"
-            class="textarea"
-            :class="[inputClasses, customClass]"
-            :maxlength="maxlength"
-            :value="computedValue === null ? undefined : computedValue"
-            v-bind="fallthroughAttrs"
-            @input="onInput"
-            @change="onChange"
-            @blur="onBlur"
-            @focus="onFocus"
-        />
+            <textarea
+                v-else
+                ref="textarea"
+                class="textarea"
+                :class="[inputClasses, customClass]"
+                :maxlength="maxlength"
+                v-model.lazy="computedValue"
+                v-bind="fallthroughAttrs"
+                @input="onInput"
+                @change="onChange"
+                @blur="onBlur"
+                @focus="onFocus"
+            />
+        </template>
+        <template v-else>
+            <input
+                v-if="type !== 'textarea'"
+                ref="input"
+                class="input"
+                :class="[inputClasses, customClass]"
+                :type="newType"
+                :autocomplete="newAutocomplete"
+                :maxlength="maxlength"
+                v-model="computedValue"
+                v-bind="fallthroughAttrs"
+                @input="onInput"
+                @change="onChange"
+                @blur="onBlur"
+                @focus="onFocus"
+            >
+
+            <textarea
+                v-else
+                ref="textarea"
+                class="textarea"
+                :class="[inputClasses, customClass]"
+                :maxlength="maxlength"
+                v-model="computedValue"
+                v-bind="fallthroughAttrs"
+                @input="onInput"
+                @change="onChange"
+                @blur="onBlur"
+                @focus="onFocus"
+            />
+        </template>
 
         <b-icon
             v-if="icon"
@@ -282,22 +315,19 @@ export default defineComponent({
             }
         },
 
-        onInput(event: Event) {
+        onInput() {
             if (!this.lazy) {
-                const value = (event.target as HTMLInputElement | HTMLTextAreaElement).value
-                this.updateValue(value)
+                this.revalidate()
             }
         },
 
-        onChange(event: Event) {
+        onChange() {
             if (this.lazy) {
-                const value = (event.target as HTMLInputElement | HTMLTextAreaElement).value
-                this.updateValue(value)
+                this.revalidate()
             }
         },
 
-        updateValue(value: string | number | undefined) {
-            this.computedValue = value
+        revalidate() {
             !this.isValid && this.checkHtml5Validity()
         }
     }

